src/errorpage.cc

```diff
@@ -1248,25 +1328,127 @@ ErrorState::BuildContent()
 MemBuf *ErrorState::ConvertText(const char *text, bool allowRecursion)
 {
     MemBuf *content = new MemBuf;
+    content->init();
+    convertAndWriteTo(text, *content, false, allowRecursion);
+    content->terminate();
+    assert((size_t)content->contentSize() == strlen(content->content()));
+    return content;
+}
+
+const char *
+ErrorState::NextCode(const char *p)
+{
+    assert(p);
+    do {
+        while (*p && *p != '%' && *p != '@') ++p;
+        if (*p == '@' && LogFormatStart.cmp(p, LogFormatStart.length()) != 0) {
+            ++p;
+            continue;
+        }
+        return p;
+    } while (1);
+}
+
+void
+ErrorState::convertAndWriteTo(const char *text, MemBuf &result, bool building_deny_info_url, bool allowRecursion)
+{
     const char *p;
     const char *m = text;
     assert(m);
-    content->init();
 
-    while ((p = strchr(m, '%'))) {
-        content->append(m, p - m);  /* copy */
-        const char *t = Convert(*++p, false, allowRecursion);   /* convert */
-        content->appendf("%s", t);   /* copy */
-        m = p + 1;          /* advance */
+    while (*(p = NextCode(m))) {
+        result.append(m, p - m);
+        if (*p == '%') {
+            const char *t = convert(++p, building_deny_info_url, allowRecursion);
+            result.appendf("%s", t);
+            m = p + 1;
+        } else {
+            assert(*p == '@');
+            handleLogFormat(p, result);
+            m = p;
+        }
     }
 
     if (*m)
-        content->appendf("%s", m);   /* copy tail */
+        result.appendf("%s", m);   /* copy tail */
+}
 
-    content->terminate();
+void
+ErrorState::handleLogFormat(const char *&start, MemBuf &result)
+{
+    assert(LogFormatStart.cmp(start, LogFormatStart.length()) == 0);
+
+    const char *p = start + LogFormatStart.length();
+    if (int tokenBytes = Format::Format::AssembleToken(p, result, al)) {
+        p += tokenBytes;
+        if (*p != '}') {
+            if (errorHandler_)
+                errorHandler_->handleError(SBuf("unterminated @Squid{"));
+            start = p;
+        } else
+            start = p + 1;
+        return;
+    }
 
-    assert((size_t)content->contentSize() == strlen(content->content()));
+    if (errorHandler_)
+        errorHandler_->handleError(ToSBuf("Cannot parse @Squid{logformat} in error page template near '", SBuf(start).substr(0, 100), "'"));
+
+    // Try recover to continue parsing
+    // The p points after '@Squid{'. Try to find the '}'.
+    if (const char *endOfComment = strchr(p, '}'))
+        p = endOfComment + 1;
+    else
+        while (*p == '%') ++p; // just skip the first '%' after '@Squid{'
```

Comment:
> If we should not recover

Please note that I did not suggest that we should not recover. I only said that we should not recover _if_ recovery can lead to wrong macro expansions. I hope you can find the conditions where wrong macro expansions are possible and stop recovery in those cases. If those conditions apply to _all_ cases or if it is difficult to reliably detect any significant safe recovery cases, then we will not recover at all.

Reply:
Current behaviour is the following:
 Assume that after @Squid{code but code is not a valid logformat code. Then 
 - Try to find the '}' termination char and if found start parsing the error page after this point.
 - If '}' termination char is not found then skips the '%' after the @Squid{ and continue parsing.

In the second case I am assuming that in most cases there is a @Squid{%code the %code is not logformat code and the user forgot to terminate the @Squid{.
In this case I preferred to not parse the %code after @Squid, because in most cases a %logformat code will be handled as %c errorpage code. But still we may have unhandled cases for example when there is a space after "@Squid{"

Maybe in this case where both %code is not logformat code and there is not a '}' termination char we can stop parsing the template. 
We are still recovering correctly in the case a termination '}' found.

Reply:
I suggest a simpler (and stricter) algorithm: If we fail to parse `@squid{...` internals or if we do not find the closing `}` bracket after successfully parsing  `@squid{...` internals, then we stop interpreting the template (at the previously found `@squid` boundary).

Are there any significant advantages in recovering after `@squid` failures? I do not see them, and I am worried that misinterpreting `@squid` boundaries may lead to wrong single-letter %code expansion.

It is often difficult to increase parsing restrictions. It is usually easy to relax them. If we discover that the above strict algorithm is too restrictive, we can always add some forgiving recovery code without running into backward compatibility problems.

Reply:
Here is an example of a trivial mistake that would result in wrong `%h` expansion if we try to recover by looking for a closing curly brace. The author thought that `@Squid{}` sequences support multiple %logformat codes and wrote this:

```
    text1 @Squid{%{foo}note %{bar}note %http::ru} text2
```

Token::parse() will stop after `%{foo}note`, right? Since that is not followed by `}, we will be in recovery mode. Recovery code would find `}` in `%{bar}note` and think that `%http` contains an old %h macro.

Reply:
ok.
The latest commit b048aada63f4b23158b52686fdcaf50226ad4d7a stops interpreting template after a missing '}' or when fails to parse @squid{ internals.